### ISSUE TITLE
Parallelize TextCorpus

### ISSUE THREAD

**Issue by BernierCR** [zalandoresearch/flair#1018]
TextCorpus is a necessary and slow step to doing anything with Flair. It does not seem to support any multi threading, but at the surface at least, seems like an easily parallelizable task. 

1 GPU comes with like 16 vCPU and 128GB of RAM, details vary based on cloud/instance type, but still. The vCPUs/RAM are practically free compared to the GPU cost!. Let's use everything we got. Is there a relatively easy way to achieve this? Loading 1GB of text data with just one thread takes some time, probably at least 10 minutes. And if you tried to load 10GB of data, it'd take hours if it doesn't just crash. 

And before I could think about doing something like figure our how to run Flair on Spark, I'd need a time estimate for how long it'd take me to do that. I don't know the scope of the changes necessary to pull this off. 

The cost of our time is a lot more expensive then the cost of even relatively massive instances. Your company would prefer to give you a bigger instance then pay you to stare at the screen waiting for things to finish, right? Let's reduce the single threaded parts as much as we can. 

Thanks.

**Comment by alanakbik** [comment]
Hello @BernierCR the `TextCorpus` currently holds `TextDataset` objects which implement the `Dataset` interface and so used in the `DataLoader` when training, meaning that it automatically parallelizes across multiple threads the loading of the next training splits.

If you split your training dataset across multple files there should be no waiting on loading text data after the first split and training LMs should reach 100% GPU utilization. So insteaf of 1 GB of text data you could do 20 splits with 50 MB or so and set the `num_workers` on the `LanguageModelTrainer.train()` to the number of CPUs you have available.

**Comment by BernierCR** [comment]
Hello again @alanakbik. Thanks for the response and the hard work. I so have my training data split, and I do have num_workers set to a nice value, and I do get 100% utilization during training. So thanks for that. 

I was referring to before training. You know the stage where it says stuff like Read 345343 lines, and shuffling. This part does not seem to be parallel. In the big picture, this is a short amount of time relative to the GPU training, but if you are waiting for it to finish so you ca be sure that the training has actually started, it can feel much longer. 

It could be trivially easy to fix this or hard. I am not familiar enough with the code base to be able to make difficulty estimates. If it's really easy, just tell me how, I'll do it. If it's hard, NBD. Thanks.

**Comment by alanakbik** [comment]
Ah yes you're right: there is to CPU parallelization in this part and parallelizing could make sense. I unfortunately haven't worked much with this type of parallelization in Python so I'm not sure how easy or hard it is to implement. At some point, I plan on reading up on this because we have CPU bottlenecks in some parts of the code where a good multi-CPU solution might really help!

**Comment by quancq** [comment]
@alanakbik "Read xxx lines and shuffle" before training is read valid and test file? If library support multiple files (as train files) then I think can use multi processing in python to read efficent way multiple file.

**Comment by richardliaw** [comment]
Maybe consider using the core [Ray APIs](https://ray.readthedocs.io/en/latest/walkthrough.html) for this? Ray provides a much cleaner API for python multiprocessing and seamlessly scales across machines - not that this is not Tune, which is a separate library packaged as part of Ray.

**Comment by ndolev** [comment]
Hi,

I have the same problem with a large corpus. I can succesfully parralelize by splitting the corpus into chunks and then doing something like this with Dask:

```{Python}
def delayedfunc(corpuspath, dictionary = dictionary):
    corpus = TextCorpus(corpuspath,
                        dictionary = dictionary,
                        forward = True,
                        character_level=True)
    return(corpus)

lazy_results = []

corpuspaths = ["/home/ubuntu/corpi/"+d for d in os.listdir("/home/ubuntu/corpi/")]
for corpuspath in corpuspaths:
    lazy_res = dask.delayed(delayedfunc)(corpuspath)
    lazy_results.append(lazy_res)
corpi = dask.compute(*lazy_results)
```
My problem is the next step. I expected to be able to do:
```{Python}
multi_corpus = MultiCorpus(list(corpi))
```
My corpus is of the format:
```
corpus/
corpus/train/
corpus/train/train_split_1
corpus/train/train_split_2
corpus/train/...
corpus/train/train_split_X
corpus/test.txt
corpus/valid.txt
```
But it seems multi-corpus is expecting a ColumnCorpus rather than a standard corpus since I get the error:
```{Python}
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-3-6b411bd01e81> in <module>
      2 # get your corpus, process forward and at the character level
      3 
----> 4 corpus = MultiCorpus(list(corpi))
      5 # corpus = TextCorpus('/home/ubuntu/corpus',
      6 #                     dictionary,

~/anaconda3/envs/Noah_PySpark37/lib/python3.7/site-packages/flair/data.py in __init__(self, corpora, name)
   1266         super(MultiCorpus, self).__init__(
   1267             ConcatDataset([corpus.train for corpus in self.corpora]),
-> 1268             ConcatDataset([corpus.dev for corpus in self.corpora]),
   1269             ConcatDataset([corpus.test for corpus in self.corpora]),
   1270             name=name,

~/anaconda3/envs/Noah_PySpark37/lib/python3.7/site-packages/flair/data.py in <listcomp>(.0)
   1266         super(MultiCorpus, self).__init__(
   1267             ConcatDataset([corpus.train for corpus in self.corpora]),
-> 1268             ConcatDataset([corpus.dev for corpus in self.corpora]),
   1269             ConcatDataset([corpus.test for corpus in self.corpora]),
   1270             name=name,

AttributeError: 'TextCorpus' object has no attribute 'dev'
```

**Comment by alanakbik** [comment]
@ndolev ah yes, the `MultiCorpus` object currently only works for labeled datasets, like `ColumnCorpus` or `ClassificationCorpus`. Is it possible to put all training files into the same 
corpus/train/ folder? In that case, you don't need a `MultiCorpus` and the `TextCorpus` will parallelize along splits automatically.

**Comment by ndolev** [comment]
@alanakbik Thanks!

Unfortunately the TextCorpus does not parallelize in my experience. I have partitioned the data into train_split_x files. 

I can construct a corpus using the Sentence API. And distribute like so (getting around GIL):

```{Python}
import dask.bag as db
import os
from flair.data import Sentence, Corpus
from flair.datasets import SentenceDataset

train = db.read_text(os.path.join(corpuspath,'train')+'/part*.txt')
train = train.map(Sentence).compute()
train_dataset = SentenceDataset(train)
.
.
.
corpus = Corpus(train_dataset, dev_dataset, test_dataset)
```

Not sure if the resultant corpus is the same as when I run TextCorpus - it's not clear to me what (if any) are the differences.

**Comment by alanakbik** [comment]
There are two types of corpora in Flair: one to train downstream tasks and one to train a language model. They are entirely different, so they cannot be mixed and matched: 

- `TextCorpus` is completely different from all others - it is an object specifically used to train a `LanguageModel`. So it cannot be used interchangeably with `Corpus`, `MultiCorpus`, `SentenceDataset` etc. It is optimized specifically for training language models fast, so it does not use `Sentence` objects. If your goal is to train a language model, you presently have to use this object.

- All other corpora extend the `Corpus` base class, including `MultiCorpus`, `ColumnCorpus` and `ClassificationCorpus`. They can also be initialized using `SentenceDataset` like you did. However, this type of corpus cannot be used to train a language model. It can be used to train sequence labelers, classifiers etc.

**Comment by ndolev** [comment]
Hi @alanakbik ,

Thank you so much for that explanation - it's much clearer to me now. I will try to distribute the base TextCorpus class and post my results here. 

Are there any plans on extending MultiCorpus to support TextCorpi? It's just if it supported it, distributing the TextCorpus would be trivial. 

Thanks again!

**Comment by ndolev** [comment]
Hi,

So here is what I managed in the meantime:

1- Take a properly arranged corpus directory and split it into chunks
```{Python}
import os
import numpy as np
from shutil import copyfile
files = [f for f in os.listdir('/home/ubuntu/corpus/train') if 'train_split' in f]
idx = np.argsort([int(f.split('_')[-1].split('.txt')[0]) for f in files])
files = [files[i] for i in idx]
n = 45  %Number of files per chunk
counter = -1

def copyfiles(file, counter):
    for ii,f in enumerate(file):
        if ii == 1: # make the first file the test (in any case, it's already been randomized)
            copyfile("/home/ubuntu/corpus/train/"+f, "/home/ubuntu/corpi/corpus"+str(counter)+'/test.txt')
        elif ii==2: # make the second file the validation (in any case, it's already been randomized)
            copyfile("/home/ubuntu/corpus/train/"+f, "/home/ubuntu/corpi/corpus"+str(counter)+'/valid.txt')
        else:
            copyfile("/home/ubuntu/corpus/train/"+f, "/home/ubuntu/corpi/corpus"+str(counter)+'/train/'+f)      

lazycopy = []            
for i in range(0, len(files), n):
    counter+=1
    file = files[i:i+n]
    lazycopy.append(dask.delayed(copyfiles)(file,counter))

res = dask.compute(*lazycopy)
```
2 - Create a corpus for each chunk

```{Python}
language_model = FlairEmbeddings('multi-forward').lm
dictionary = language_model.dictionary

def createcorp(path,dictionary = dictionary,
               is_forward_lm = True):
    
    corpus = TextCorpus(path, 
                    dictionary, 
                    is_forward_lm, 
                    character_level=True)
    return(corpus)

lazy_res = [dask.delayed(createcorp)(os.path.join('/home/ubuntu/corpi/',f)) for f in os.listdir('/home/ubuntu/corpi/')]
corpi = dask.compute(*lazy_res)
```
3 - Train sequentially (but each time in parallel) on each corpus:

```{Python}
resultspath = "/home/ubuntu/temp_results"

language_model = FlairEmbeddings('multi-forward').lm

for corpus in corpi:
    trainer = LanguageModelTrainer(language_model, corpus)
    trainer.train(resultspath,
                  sequence_length=100,
                  mini_batch_size=100,
                  learning_rate=20,
                  patience=10, 
                  num_workers=95,
                  checkpoint=True)
```
If there were some way to merge the chunked TextCorpus, this could be an easy solution...

**Comment by alanakbik** [comment]
Thanks for sharing this - how does part 3 run in parallel, i.e. are separate processes training the LM over different splits of the corpus? Are they all separately making weight updates to the LM? 

Also, to better understand the problem: Why would parallelization be trivial if `TextCorpus` were an extension of the Corpus base class?

**Comment by ndolev** [comment]
> Thanks for sharing this - how does part 3 run in parallel, i.e. are separate processes training the LM over different splits of the corpus? Are they all separately making weight updates to the LM?

No, this is running parallel training sessions serially (each training happens in parallel or on a GPU). I realize I forgot a line. I should have wrote:

```{Python}
for corpus in corpi:
    trainer = LanguageModelTrainer(language_model, corpus)
    
#     trainer.model = _CustomDataParallel(trainer.model)
    
    trainer.train(resultspath,
                  sequence_length=256,
                  mini_batch_size=100,
                  learning_rate=20,
                  patience=10, 
                  num_workers=ncores,
                  checkpoint=True)
    language_model = FlairEmbeddings(os.path.join(resultspath,'best-lm.pt'))
```
> 
> Also, to better understand the problem: Why would parallelization be trivial if `TextCorpus` were an extension of the Corpus base class?

If I could combine the chunked corpi with MultiCorpus, then I could run the training without iterating over the chunks (serially). And if we solved #499 (where it seems like a very small change is needed), then we could run the whole LM generation distributed.

**Comment by alanakbik** [comment]
Sounds good, though I am not sure I understand the details. Would you be interested in coding this up in a branch so we can test? If it works, we can do a PR to the main project.
